tests/mod.rs

```diff
@@ -0,0 +1,133 @@
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+
+use simple_irc::Message;
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct TestAtoms {
+    #[serde(default)]
+    tags: BTreeMap<String, String>,
+    source: Option<String>,
+    verb: String,
+    #[serde(default)]
+    params: Vec<String>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct MsgSplitTest {
+    input: String,
+    atoms: TestAtoms,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct MsgSplitTests {
+    tests: Vec<MsgSplitTest>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct MsgJoinTest {
+    desc: String,
+    matches: Vec<String>,
+    atoms: TestAtoms,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct MsgJoinTests {
+    tests: Vec<MsgJoinTest>,
+}
+
+#[test]
+fn test_msg_split() {
+    let msg_split_test_data = include_str!("external/parser-tests/tests/msg-split.yaml");
+    let tests = serde_yaml::from_str::<MsgSplitTests>(msg_split_test_data).unwrap();
+
+    for test in tests.tests {
+        println!("Trying {}", &test.input);
+
+        let res = Message::try_from(&test.input[..]);
+
+        // Ensure all messages parse into something
+        assert!(
+            !res.is_err(),
+            "msg failed: \"{}\", err {}",
+            &test.input,
+            res.unwrap_err()
+        );
+
+        let msg = res.unwrap();
+
+        let mut test_tags = test.atoms.tags.clone();
+        let mut msg_tags = msg.tags.clone();
+
+        // Loop through all the test tags and make sure they were there.
+        for (key, value) in test_tags.clone() {
+            assert_eq!(
+                &value,
+                msg_tags.get(key.as_str()).unwrap(),
+                "Mismatched value for key {}",
+                key.as_str()
+            );
+            test_tags.remove(&key);
+            msg_tags.remove(&key[..]);
+        }
```

Comment:
You don't actually need to remove from test_tags every time you check a key (so no need to clone).  And you can use 'remove' instead of 'get' in your assert:

```
for (key, value) in test_tags {
    assert_eq!(
        *value,
        msg_tags.remove(key).unwrap(),
        "Mismatched value for key {}",
        key
    );
}
```

Reply:
Ah, cool